core/notification/src/main/java/edu/stanford/spezi/core/notification/fcm/DeviceRegistrationService.kt

```diff
@@ -0,0 +1,115 @@
+package edu.stanford.spezi.core.notification.fcm
+
+import android.content.Context
+import android.os.Build
+import com.google.firebase.functions.FirebaseFunctions
+import dagger.hilt.android.qualifiers.ApplicationContext
+import edu.stanford.spezi.core.logging.speziLogger
+import java.util.Locale
+import java.util.TimeZone
+import javax.inject.Inject
+
+/**
+ * Service to register the device with the server.
+ */
+class DeviceRegistrationService @Inject constructor(
+    @ApplicationContext private val context: Context,
+    private val functions: FirebaseFunctions,
+    private val messageTokenService: MessageTokenService,
+) {
+
+    private val logger by speziLogger()
+
+    /**
+     * Registers the device with the server.
+     */
+    suspend fun registerDevice() {
+        val notificationToken = getNotificationToken().getOrNull() ?: run {
+            logger.e { "Notification token is null" }
+            return
+        }
+        val platform = PLATFORM_ANDROID
+        val osVersion = Build.VERSION.RELEASE
+        val appVersion = getAppVersion()
+        val appBuild = getAppBuild()
+        val language = Locale.getDefault().toLanguageTag()
+        val timeZone = TimeZone.getDefault().id
+
+        val deviceInfo = DeviceInfo(
+            notificationToken = notificationToken,
+            platform = platform,
+            osVersion = osVersion,
+            appVersion = appVersion,
+            appBuild = appBuild,
+            language = language,
+            timeZone = timeZone
+        )
+
+        sendDeviceInfoToServer(deviceInfo)
+    }
+
+    private suspend fun getNotificationToken(): Result<String?> {
+        return runCatching {
+            messageTokenService.getNotificationToken()
+        }.onFailure { e ->
+            logger.e(e) { "Error getting notification token" }
+        }
+    }
+
+    private fun getAppVersion(): String {
```

Comment:
Would update to one single `private fun getPackageInfo() = context.packageManager.getPackageInfo(context.packageName, 0)` and use `packageInfo.versionName` and `packageInfo.versionCode` above

Reply:
What I meant here was to invoke `getPackageInfo()` just once:

```kotlin
val packageInfo = getPackageInfo()
...
val deviceInfo = DeviceInfo(
            notificationToken = token,
            platform = platform,
            osVersion = osVersion,
            appVersion = packageInfo.versionName,
            appBuild = packageInfo.versionCode.toString(),
            language = language,
            timeZone = timeZone
        )
```